torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+#pragma once
+#include <c10/core/impl/TorchDispatchModeTLS.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/utils/python_stub.h>
+#include <torch/csrc/utils/torch_dispatch_mode.h>
+#include <iostream>
+#include <typeindex>
+#include <vector>
+
+// see [Note: Compiled Autograd]
+
+namespace torch {
+namespace autograd {
+namespace generated {
+struct TypeAndSize;
+}
+
+struct SizeInput {
+  enum DynType : uint8_t { STATIC = 0, DYNAMIC = 1 };
+  SizeInput(DynType dt, int64_t v) : dyn_type(dt), value(v) {}
+  DynType dyn_type;
+  int64_t value;
+};
+
+struct CacheKeyBuffer {
+  CacheKeyBuffer() : data(nullptr) {}
+  CacheKeyBuffer(const uint8_t* key, uint16_t len) : data(new uint8_t[len]) {
+    std::memcpy(data, key, len);
+  }
+  CacheKeyBuffer(CacheKeyBuffer&& other)
+      : data(std::exchange(other.data, nullptr)) {}
+  ~CacheKeyBuffer() {
+    if (data != nullptr) {
+      delete[] data;
+    }
+  }
+  CacheKeyBuffer(const CacheKeyBuffer& other) = delete;
+  CacheKeyBuffer& operator=(const CacheKeyBuffer&) = delete;
+
+  uint8_t* data;
+};
+
+struct CacheKey {
+  // Key to find the next node in the shadow graph.  We use C++ RTTI for the
+  // type of the node (ntype), then a key generated with a visitor pattern.
+  CacheKey(const std::type_index& ntype, const uint8_t* key, uint16_t len)
+      : node_type(ntype), key_size(len), key(key) {}
+
+  bool operator<(const CacheKey& other) const {
+    if (node_type != other.node_type) {
+      return node_type < other.node_type;
+    }
+    if (key_size != other.key_size) {
+      return key_size < other.key_size;
+    }
+    return std::memcmp(key, other.key, key_size) < 0;
+  }
+
+  bool operator==(const CacheKey& other) const {
+    return node_type == other.node_type && key_size == other.key_size &&
+        std::memcmp(key, other.key, key_size) == 0;
+  }
+
+  size_t hash() const {
+    // don't bother hashing the key data, common case 1 cache entry per node
+    return std::hash<std::type_index>()(node_type) ^ key_size;
+  }
+
+  std::type_index node_type;
+  uint16_t key_size;
+  const uint8_t* key;
+};
+
+struct OutputRef {
+  OutputRef() : node_id(-1), index(-1) {}
+  OutputRef(int node_, int index_) : node_id(node_), index(index_) {}
+
+  bool is_set() const {
+    return this->node_id >= 0;
+  }
+
+  int node_id;
+  int index;
+};
+
+struct NodeCall {
+  NodeCall(const std::shared_ptr<Node>& node_)
+      : node(node_), input_refs(node_->num_inputs()) {}
+
+  NodeCall(const std::shared_ptr<Node>& node_, int num_inputs_)
+      : node(node_), input_refs(num_inputs_) {}
```

Comment:
Is there an invariant that `node_->num_inputs() == num_inputs_`?

Reply:
No, we violate that when we create `ImplicitAdd()` nodes for multiple edges going into the same node.  We never setup the metadata to populate `node_->num_inputs()` in those cases.